expfmt/openmetrics_parse.go

```diff
@@ -0,0 +1,363 @@
+// frequent use of this method, which is sample ingestion in the Prometheus
+// server. However, for presentation purposes, you might want to sort the
+// metrics, and in some cases, you must sort the labels, e.g. for consumption by
+// the metric family injection hook of the Prometheus registry.
+//
+// Summaries and histograms are rather special beasts. You would probably not
+// use them in the simple openmetrics format anyway. This method can deal with
+// summaries and histograms if they are presented in exactly the way the
+// openmetrics.Create function creates them.
+//
+// This method must not be called concurrently. If you want to parse different
+// input concurrently, instantiate a separate Parser for each goroutine.
+func (p *OpenMetricsParser) OpenMetricsToMetricFamilies(in io.Reader) (map[string]*dto.MetricFamily, error) {
+	p.reset(in)
+	p.currentIsEOF = false
+	for nextState := p.startOfLine; nextState != nil; nextState = nextState() {
+		// Magic happens here...
+	}
+	// If p.err is io.EOF now, we have run into a premature end of the input
+	// stream. Turn this error into something nicer and more
+	// meaningful. (io.EOF is often used as a signal for the legitimate end
+	// of an input stream.)
+	if p.err != nil && errors.Is(p.err, io.EOF) {
+		p.parseError("unexpected end of input stream")
+	}
+	return p.metricFamiliesByName, p.err
+}
+
+func (p *OpenMetricsParser) reset(in io.Reader) {
+	p.metricFamiliesByName = map[string]*dto.MetricFamily{}
+	if p.buf == nil {
+		p.buf = bufio.NewReader(in)
+	} else {
+		p.buf.Reset(in)
+	}
+	p.err = nil
+	p.lineCount = 0
+}
+
+// startOfLine represents the state where the next byte read from p.buf is the
+// start of a line (or whitespace leading up to it).
+func (p *OpenMetricsParser) startOfLine() stateFn {
+	p.lineCount++
+	p.currentByte, p.err = p.buf.ReadByte()
+	if p.err != nil {
+		// This is the only place that we expect to see io.EOF,
+		// which is not an error but the signal that we are done.
+		// Any other error that happens to align with the start of
+		// a line is still an error.
+		if errors.Is(p.err, io.EOF) {
+			if p.currentIsEOF {
+				p.err = nil
+			} else {
+				p.parseError("expected EOF keyword at the end")
+			}
+		}
+		return nil
+	}
+	if p.currentIsEOF {
+		p.parseError(fmt.Sprintf("unexpected line after EOF, got %q", p.currentByte))
+		return nil
+	}
+	switch p.currentByte {
+	case '#':
+		return p.startComment
+	case '\n':
+		return p.startOfLine // Empty line, start the next one.
+	}
+	if !isValidMetricNameStart(p.currentByte) {
+		p.parseError(fmt.Sprintf("%q is not a valid start token", p.currentByte))
+		return nil
+	}
+	return nil
+}
+
+// startComment represents the state where the next byte read from p.buf is the
+// start of a comment (or whitespace leading up to it).
+func (p *OpenMetricsParser) startComment() stateFn {
+	if p.skipBlankTab(); p.err != nil {
+		return nil // Unexpected end of input.
+	}
```

Comment:
I remember that we talked about this in the previous PR but I can't remember the outcome (sorry for being lazy here 😛). Does OpenMetrics allow lines starting with blanks/tabs?

Reply:
Hmmm, I've just searched OM's spec and there isn't a single mention of the words `tab` or `whitespace`. Let's keep it like this then, but I wouldn't add tests for this since we don't need to offer strong guarantees about it